### PR TITLE
copr: fix datetime fallback handler edge cases

### PR DESCRIPTION
**Pull request by iosmanthus**
Signed-off-by: Iosmanthus Teng <myosmanthustree@gmail.com>

###  What have you changed?
If there is a datetime format string (in format like `yyyy-mm-ddThh:mm:ss/yyyy-mm-dd hh:mm:ss` provided to the `Duration` parser, we need to check if the seperater is valid.

Don't assume reviewers understand the original issue.

###  What is the type of changes?

- Bugfix (a change which fixes an issue)

###  How is the PR tested?

- Unit test
- Integration test

### CODE REVIEW

File: components/tidb_query/src/codec/mysql/duration.rs [hunk 1]
```diff
@@ -146,41 +146,36 @@ mod parser {
...
-            let (rest, fraction) = fraction(input, fsp)?;
             return Ok((
-                rest,
-                [
-                    datetime.hour(),
-                    datetime.minute(),
-                    datetime.second(),
-                    fraction,
-                ],
+                &input[input.len()..],
```
Comment (lonng) [inline components/tidb_query/src/codec/mysql/duration.rs]: Do we need to `input[input.len()..]` because the rest is always `""`;

Comment (iosmanthus) [reply components/tidb_query/src/codec/mysql/duration.rs]: OK. Seems the whole function doesn't need to return an `IResult`, a simple `Result` is enough.

File: components/tidb_query/src/codec/mysql/duration.rs [hunk 2]
```diff
@@ -146,42 +146,31 @@ mod parser {
...
+            return Ok(("", datetime.convert(ctx).map_err(|_| nom::Err::Error(()))?));
         }
         let (rest, _) = anysep(rest)?;
         let (rest, _) = digit1(rest)?;
         let (rest, _) = anysep(rest)?;
         let (rest, _) = digit1(rest)?;
 
-        if rest.is_empty() {
+        let has_datetime_sep = match rest.chars().next() {
+            Some(c) if c != 'T' && c != ' ' => false,
```
Comment (andylokandy) [inline components/tidb_query/src/codec/mysql/duration.rs]: suggestion:
some(c) if c == T || c == ' ' => true
_ => false

File: components/tidb_query/src/codec/mysql/duration.rs [hunk 3]
```diff
@@ -754,6 +744,25 @@ mod tests {
         }
     }
 
+    #[test]
+    fn test_parse_datetime() {
+        let cases: Vec<(&'static [u8], i8, Option<&'static str>)> = vec![
+            (b"2010-02-12", 0, None),
+            (b"2010-02-12T12:23:34", 0, Some("12:23:34")),
```
Comment (andylokandy) [inline components/tidb_query/src/codec/mysql/duration.rs]: you can test against a case replacing T to t.

**Comment by andylokandy** [comment]
@sre-bot /merge

**Pull request merged**